File: server/app/models/user.model.ts
import mongoose, { ObjectId, Schema, model } from 'mongoose';
import { ROLE } from '../const/user.const';

export interface TUser {
    _id: string;
    name: string;
    email: string;
    password: string;
    avatar: string;
    isPublic: Boolean;
    lock: {
        status: Boolean;
        content: string;
    };
    phone?: string;
    refreshToken: string;
    role: string;
}
const UserSchema: Schema<TUser> = new Schema(
    {
        name: { type: String, max: 30, trim: true, required: true },
        email: {
            type: String,
            trim: true,
            unique: true,
            lowercase: true,
            required: true,
        },
        password: { type: String, required: true },
        avatar: {
            type: String,
            default:
                'https://firebasestorage.googleapis.com/v0/b/stadium-booking-854aa.appspot.com/o/avatar%2Favatar_user.png?alt=media&token=d64d8608-ee56-4b91-bbff-ced68a52cc06',
        },
        isPublic: { type: Boolean, default: true, required: true },
        lock: {
            status: { type: Boolean, default: false },
            content: { type: String, default: '', max: 1500 },
        },
        phone: {
            type: String,
            minLength: [10, 'no should have minimum 10 digits'],
            maxLength: [10, 'no should have maximum 10 digits'],
            match: [/\d{10}/, 'no should only have digits'],
            validator: function (v: string) {
                return v.length == 10;
            },
            unique: true,
        },
        role: { type: String, enum: ROLE, default: ROLE.user },
        refreshToken: { type: String, default: '' },
    },
    {
        timestamps: true,
        versionKey: false,
    }
);
export default model<TUser>('User', UserSchema);

File: widget_component/lib/src/models/user_model.dart
class UserModel {
  LockUser? lock;
  String? sId;
  String? name;
  String? email;
  String? avatar;
  bool? isPublic;
  String? phone;
  String? role;
  String? accessToken;
  String? refreshToken;
  int? fieldCount;

  UserModel(
      {this.lock,
      this.sId,
      this.name,
      this.email,
      this.avatar,
      this.isPublic,
      this.phone,
      this.role,
      this.fieldCount});

  UserModel.fromJson(Map<String, dynamic> json) {
    lock = json['lock'] != null ? LockUser.fromJson(json['lock']) : null;
    sId = json['_id'];
    name = json['name'];
    email = json['email'];
    avatar = json['avatar'];
    isPublic = json['isPublic'];
    phone = json['phone'];
    role = json['role'];
    fieldCount = json['fieldCount'];
    accessToken = json['accessToken'];
    refreshToken = json['refreshToken'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    if (lock != null) {
      data['lock'] = lock!.toJson();
    }
    data['_id'] = sId;
    data['name'] = name;
    data['email'] = email;
    data['avatar'] = avatar;
    data['isPublic'] = isPublic;
    data['phone'] = phone;
    data['role'] = role;
    data['fieldCount'] = fieldCount;
    data['accessToken'] = accessToken;
    data['refreshToken'] = refreshToken;
    return data;
  }
}

class LockUser {
  bool? status;
  String? content;

  LockUser({this.status, this.content});

  LockUser.fromJson(Map<String, dynamic> json) {
    status = json['status'];
    content = json['content'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    data['status'] = status;
    data['content'] = content;
    return data;
  }
}
